provider/ec2/instancetypes.go

```diff
@@ -0,0 +1,328 @@
+		}
+		instTypeParams := &ec2.DescribeInstanceTypesInput{
+			InstanceTypes: instTypeNames[0:querySize],
+		}
+		instTypeResults, err := ec2Session.DescribeInstanceTypes(instTypeParams)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, info := range instTypeResults.InstanceTypes {
+			instType := instances.InstanceType{
+				Name:       *info.InstanceType,
+				VirtType:   virtType(info),
+				Deprecated: info.CurrentGeneration == nil || !*info.CurrentGeneration,
+				Cost:       costs[*info.InstanceType],
+			}
+			if info.VCpuInfo != nil && info.VCpuInfo.DefaultVCpus != nil {
+				instType.CpuCores = uint64(*info.VCpuInfo.DefaultVCpus)
+				if info.ProcessorInfo != nil && info.ProcessorInfo.SustainedClockSpeedInGhz != nil {
+					cpupower := calculateCPUPower(
+						instType.Name,
+						info.ProcessorInfo.SustainedClockSpeedInGhz,
+						uint64(*info.VCpuInfo.DefaultVCpus))
+					instType.CpuPower = &cpupower
+				}
+			}
+			if info.MemoryInfo != nil && info.MemoryInfo.SizeInMiB != nil {
+				instType.Mem = uint64(*info.MemoryInfo.SizeInMiB)
+			}
+			if info.ProcessorInfo != nil {
+				for _, instArch := range info.ProcessorInfo.SupportedArchitectures {
+					instType.Arches = append(instType.Arches, archName(*instArch))
+				}
+			}
+			if instRegions, ok := instanceTypeRegions[instType.Name]; !ok || instRegions.Size() < len(zoneNames) {
+				instType.Deprecated = true
+			}
+			allInstanceTypes = append(allInstanceTypes, instType)
+		}
+		if len(instTypeNames) <= 100 {
+			break
+		}
+		instTypeNames = instTypeNames[querySize:]
+	}
+
+	if isVPCIDSet(e.ecfg().vpcID()) {
+		return allInstanceTypes, nil
+	}
+	hasDefaultVPC, err := e.hasDefaultVPC(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if hasDefaultVPC {
+		return allInstanceTypes, nil
+	}
+
+	// The region has no default VPC, and the user has not specified
+	// one to use. We filter out any instance types that are not
+	// supported in EC2-Classic.
+	supportedInstanceTypes := make([]instances.InstanceType, 0, len(allInstanceTypes))
+	for _, instanceType := range allInstanceTypes {
+		if !supportsClassic(instanceType.Name) {
+			continue
+		}
+		supportedInstanceTypes = append(supportedInstanceTypes, instanceType)
+	}
+	return supportedInstanceTypes, nil
+}
+
+// instanceTypeCosts queries the latest spot price for the given instance types.
+func instanceTypeCosts(ec2Session ec2iface.EC2API, instTypeNames []*string, zoneNames []string, costs map[string]uint64) error {
+	var token *string
+	spParams := &ec2.DescribeSpotPriceHistoryInput{
+		InstanceTypes: instTypeNames,
+		MaxResults:    aws.Int64(1000),
+		StartTime:     aws.Time(time.Now()),
+		Filters:       []*ec2.Filter{{Name: aws.String("availability-zone")}},
+	}
+	for _, zone := range zoneNames {
+		spParams.Filters[0].Values = append(spParams.Filters[0].Values, aws.String(zone))
+	}
+	for {
+		spParams.NextToken = token
+		priceResults, err := ec2Session.DescribeSpotPriceHistory(spParams)
```

Comment:
Shouldn't use SpotPriceHistory as this is the spot price, not the actual price. Instead use this API https://docs.aws.amazon.com/sdk-for-go/api/service/pricing/

Reply:
cost is just a guide (it's only used to break a tie when name and/or memory and/or cpu are equal)
i looked at that other api and even tried to use it but it doesn't have the filtering needed to allow it to be queried efficiently (that i could see)
spot price (or anything really) is fine for the use case here

Reply:
Considering the spot price is unstable, we need to revisit this.

Reply:
Possibly even just not using price. Instead actually parsing the instance type name, and creating a better weighting based system. T and M type instances should be chosen above everything else, then Compute and Memory optimized, then all the others after.